src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -353,24 +309,11 @@ public byte[] Http2AltSvcOriginUri
             }
 
             // Either H3 explicitly requested or secured upgraded allowed.
-            if (_http3Enabled && (request.Version.Major >= 3 || (request.VersionPolicy == HttpVersionPolicy.RequestVersionOrHigher && IsSecure)))
+            if (IsHttp3Enabled && (request.Version.Major >= 3 || (request.VersionPolicy == HttpVersionPolicy.RequestVersionOrHigher && IsSecure)))
             {
-                HttpAuthority? authority = _http3Authority;
-                // H3 is explicitly requested, assume prenegotiated H3.
-                if (request.Version.Major >= 3 && request.VersionPolicy != HttpVersionPolicy.RequestVersionOrLower)
-                {
-                    authority = authority ?? _originAuthority;
-                }
-                if (authority != null)
-                {
-                    if (IsAltSvcBlocked(authority))
-                    {
-                        return ValueTask.FromException<(HttpConnectionBase? connection, bool isNewConnection, HttpResponseMessage? failureResponse)>(
-                            new HttpRequestException(SR.Format(SR.net_http_requested_version_cannot_establish, request.Version, request.VersionPolicy, 3)));
-                    }
-
-                    return GetHttp3ConnectionAsync(request, authority, cancellationToken);
-                }
+                var result = GetHttp3ConnectionAsync(request, cancellationToken);
```

Comment:
Can you help me understand why all of these changes are needed?

My suggestion was to have just an IsHttp3Enabled or Http3.IsSupported property (the latter of which could be configured via the same mechanisms we use for other IsSupported properties) that was special-cased by platform.  If the linker could see it was false, for example, then here GetHttp3ConnectionAsync shouldn't need to have been moved and stubbed out; it would never be invoked on platforms where it wasn't supported, and the linker would be able to remove it entirely, since all calls to it are guarded by IsHttp3Enabled/Http3.IsSupported.

Reply:
The quic apis were called from more places than expected and the simplest way to automatically check that no PNSE APIs are never called is not to compile them (the analyzer does not work here). If it was only a few methods it'd be fine to put IsHttp3Enabled everywhere but how do you ensure the PNSE code is never called and no future PR break it?

Reply:
> how do you ensure the PNSE code is never called and no future PR break it?

Isn't that what the platform compatibility analyzer is for?

Reply:
It does not work here see #49278 and its method body limited so that's a lot of additional annotations to bubble it up to right IsHttp3Enabled check.